Src/DependencyCollector/Shared/Implementation/SqlClientDiagnosticSourceListener.cs

```diff
@@ -0,0 +1,539 @@
+                    }
+                }
+            }
+            else
+            {
+                telemetry.Context.Operation.Id = operationId.ToString("N");
+            }
+        }
+
+        private static void ConfigureExceptionTelemetry(DependencyTelemetry telemetry, Exception exception)
+        {
+            telemetry.Success = false;
+            telemetry.Properties["Exception"] = exception.ToInvariantString();
+
+            var sqlException = exception as SqlException;
+
+            if (sqlException != null)
+            {
+                telemetry.ResultCode = sqlException.Number.ToString(CultureInfo.InvariantCulture);
+            }
+        }
+
+        private void OnAfterCommandEvent(
+            Guid operationId,
+            string operation,
+            Guid connectionId,
+            SqlCommand command,
+            IDictionary statistics,
+            long endTimestamp)
+        {
+            var telemetry = this.CreateCommandTelemetry(operationId, command, endTimestamp);
+            
+            this.client.Track(telemetry);
+        }
+
+        private void OnCommandErrorEvent(
+            Guid operationId,
+            string operation,
+            Guid connectionId,
+            SqlCommand command,
+            Exception exception,
+            long endTimestamp)
+        {
+            var telemetry = this.CreateCommandTelemetry(operationId, command, endTimestamp);
+
+            ConfigureExceptionTelemetry(telemetry, exception);
+
+            this.client.Track(telemetry);
+        }
+
+        private DependencyTelemetry CreateCommandTelemetry(Guid operationId, SqlCommand command, long endTimestamp)
+        {
+            // Call DeriveDuration first to ensure we remove any start timestamp from operationStartTimestamps
+            var duration = this.DeriveDuration(operationId, endTimestamp);
+
+            var dependencyName = string.Empty;
+            var target = string.Empty;
+
+            if (command.Connection != null)
+            {
+                target = string.Join(" | ", command.Connection.DataSource, command.Connection.Database);
+
+                var commandName = command.CommandType == CommandType.StoredProcedure
+                    ? command.CommandText
+                    : string.Empty;
+
+                dependencyName = string.IsNullOrEmpty(commandName)
+                    ? string.Join(" | ", command.Connection.DataSource, command.Connection.Database)
+                    : string.Join(" | ", command.Connection.DataSource, command.Connection.Database, commandName);
+            }
+
+            var telemetry = new DependencyTelemetry()
+            {
+                Id = operationId.ToString("N"),
+                Name = dependencyName,
+                Type = RemoteDependencyConstants.SQL,
+                Target = target,
+                Data = command.CommandText,
+                Duration = duration,
+                Timestamp = DateTimeOffset.UtcNow - duration,
```

Comment:
We anyway store the precise timestamp when operation started, so we should just use that, instead of deriving, which wont be as accurate. (There is a UI showing dependency timeline, which could get affected if start time is not as precise as possible, esp. for fast calls.)

Reply:
We are not storing the start timestamp, but the ticks value returned by Stopwatch.GetTimestamp().

Reply:
ok

Reply:
AFAIK, the clock resolution in `UtcNow` is 10-15 ms; adding time it takes the code to execute between `DeriveDuration` and `UtcNow` (especially if this code may yield execution to the other task) we can get an interval big enough to put the dependency into the wrong spot on the timeline relative to the parent request. We may consider use higher precision timers and/or make sure we calculate duration/start as close as possible (still prone to thread tricks...).

Reply:
Does this problem go away if we use CacheBasedOperationHolder? Presumably, we can then just record the timestamp when we get the begin event.

Reply:
Yes i think we store the start timestamp inside the telemetry object kept inside the cache.
Check this:
https://github.com/Microsoft/ApplicationInsights-dotnet-server/blob/develop/Src/DependencyCollector/Shared/Implementation/FrameworkSqlProcessing.cs